File: transferOut/Binary_tree_iterator.cpp
#include<stack>
#include<iostream>
#include<memory>

using namespace std;
struct TreeNode{
int val;
TreeNode* left;
TreeNode* right;
TreeNode():val(0), left(NULL), right(NULL) {};
};

struct BSTIterator { 
  TreeNode* root;

  BSTIterator(TreeNode* node){ root = node; };
    stack<TreeNode*>st;
    int next() { 
        while(root){
            st.push(root); 
           root = root -> left;
        };
     
    TreeNode* value = st.top();
        st.pop(); 
        root =  value->right;
        int val = value->val;
        delete value;
        return val;
    };

    bool hasNext() {
         return root || st.size();
    };
};

TreeNode * makeBinary(int arr[], int hi, int lo = 0, TreeNode* BST = new TreeNode()){
    if(lo > hi)return NULL;
     int mid = lo + (hi - lo >> 1);
     BST->val = arr[mid];
     BST->left =  makeBinary(arr, mid - 1, lo, new TreeNode());
     BST->right = makeBinary(arr, hi, mid + 1, new TreeNode());
     return BST;
}



int main(){

int arr[9] = {3,4,5,6,7,8,9,10,11};
int l = 9;
TreeNode *root = makeBinary(arr, l - 1);
BSTIterator*  Iterator = new BSTIterator(root);
while(Iterator->hasNext()){
cout << Iterator->next() << "\n"; 
    }
    
    delete Iterator;
    return 0;
}

File: Binary_Tree_iterator.js




function BSTIterator(root) {
  const stack = []

  return {
    next,
    hasNext
  }

  function next() {
    while (root) {
      stack.push(root)
      root = root.left;
    }
    const val = stack.pop();
    root = val.right;
    return val.val;
  }

  function hasNext() {
    return root || stack.length;
  }

}


function ToBinary(arr){

return bind()

function bind(lo = 0, hi = arr.length - 1, node = {}){
if(lo > hi)return null;
const mid = lo + (hi - lo >> 1);
node.val = arr[mid]; 
node.left = bind(lo, mid - 1, {});
node.right = bind(mid + 1, hi, {});
return node;
  }
}




const arr = [2,7,3,15,20];

const binaryTree = ToBinary(arr);

let iterator = new BSTIterator(binaryTree);

let a = iterator/*?*/



a.next() /*?*/
a.next()/*?*/
iterator.hasNext() /*?*/
a.next() /*?*/
iterator.next() /*?*/
iterator.hasNext() /*?*/
iterator.next() /*?*/

iterator.hasNext() /*?*/









